compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -186,6 +198,11 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
             }
         }
 
+        if alloc.get_relocations().is_empty() {
+            // There are no refs or relocations in this allocation, we can skip the interning walk.
+            return Ok(());
+        }
```

Comment:
To not change the existing process too much (keeping the existing order of checks: ZST, then UnsafeCell) and avoid the optimization if any interior mutability happens in general: we thought the complexity of handling both would not be worth it just for this rare combination. 

Is that description accurate @oli-obk ?

Reply:
Well, but you did add some new short-circuiting that happens before the UnsafeCell path. And The UnsafeCell path is also only relevant for the references we encounter in there.

In the end, what all your logic here should do is compute a single `bool`: "can we skip this?". I suggest to structure the code accordingly (e.g. in a helper closure, so that you can easily short-circuit that computation). That `bool` should be `false` if we cannot determine the size. But if we decide that we can indeed skip, then we should do that before even testing for UnsafeCell.

Reply:
Are you referring to the extern types early return as "new short circuiting" ? I’m not sure this has changed in practice, the UnsafeCell ty would not match in this situation, right ?

Reply:
I think it might match, in situations like `UnsafeCell<(&T, ExternType)>`.

Reply:
To err on the safer side and to cleanup the function, I've done the refactor you suggest, in https://github.com/rust-lang/rust/pull/97585/commits/4bcb6118ff6fbe091c1cc5d7ea306bdc9e39796c